presto-common/src/main/java/com/facebook/presto/common/type/TypeUtils.java

```diff
@@ -212,4 +213,52 @@ static <V> Map<String, V> normalizeEnumMap(Map<String, V> entries)
         return entries.entrySet().stream()
                 .collect(toMap(e -> e.getKey().toUpperCase(ENGLISH), Map.Entry::getValue));
     }
+
+    /**
+     * For our definitions of double and real, Nan=NaN is true
+     * NaN is greater than all other values, and +0=-0 is true.
+     * the below functions enforce that definition
+     */
+    public static boolean doubleEquals(double a, double b)
+    {
+        // the first check ensures +0 == -0 is true. the second ensures that NaN == NaN is true
+        // for all other cases a == b and doubleToLongBits(a) == doubleToLongBits(b) will return
+        // the same result
+        return a == b || doubleToLongBits(a) == doubleToLongBits(b);
+    }
+
+    public static long doubleHashCode(double value)
+    {
+        // canonicalize +0 and -0 to a single value
+        value = value == -0 ? 0 : value;
+        return AbstractLongType.hash(doubleToLongBits(value));
+    }
+
+    public static int doubleCompare(double a, double b)
+    {
+        // these three ifs can only be true if neither value is NaN
+        if (a < b) {
+            return -1;
+        }
+        if (a > b) {
+            return 1;
+        }
+        // this check ensure doubleCompare(+0, -0) will return 0
+        // if we just did doubleToLongBits comparison, then they
+        // would not compare as equal
+        if (a == b) {
+            return 0;
+        }
+
+        // this ensures that doubleCompare(NaN, NaN) will return 0
+        long aBits = doubleToLongBits(a);
+        long bBits = doubleToLongBits(b);
+        if (aBits < bBits) {
+            return -1;
+        }
+        if (aBits > bBits) {
+            return 1;
+        }
+        return 0;
```

Comment:
In theory, isn't it possible for multiple NaNs to have separate values when represented as bits?  Supposing this came from a datasource and not from Presto itself.  If so, would this work?

Reply:
doubleToLongBits coerces all the nans to the same representation.  There's a different function Double.doubelToRawLongBits() that doesn't do that.

Reply:
I added a comment in the code and some tests for this.